cpp/monitoring/gcm/exporter.cc

```diff
@@ -0,0 +1,264 @@
+}
+
+
+void GCMExporter::RefreshCredentials() {
+  VLOG(1) << "Refreshing GCM credentials...";
+  UrlFetcher::Request req(
+      (URL(FLAGS_google_compute_metadata_url + "/" +
+           FLAGS_google_compute_monitoring_service_account + "/token")));
+  req.headers.insert(make_pair("Metadata-Flavor", "Google"));
+
+  UrlFetcher::Response resp;
+  SyncTask task(executor_);
+  fetcher_->Fetch(req, &resp, task.task());
+  task.Wait();
+
+  // TODO(alcutter): Handle this better
+  CHECK_EQ(util::Status::OK, task.status());
+  CHECK_EQ(200, resp.status_code);
+  token_refreshed_at_ = system_clock::now();
+
+  JsonObject reply(resp.body);
+  CHECK(reply.Ok()) << "Failed to parse metadata JSON:\n" << resp.body;
+  JsonString bearer(reply, "access_token");
+  CHECK(bearer.Ok());
+  bearer_token_ = bearer.Value();
+
+  VLOG(1) << "GCM credentials refreshed";
+}
+
+
+namespace {
+
+
+void AddLabelDescription(const string& key, const string& desc,
+                         JsonArray* labels) {
+  JsonObject l;
+  l.Add("key", kCloudPrefix + key);
+  l.Add("description", desc);
+  CHECK_NOTNULL(labels)->Add(&l);
+}
+
+
+}  // namespace
+
+
+void GCMExporter::CreateMetrics() const {
+  const std::set<const Metric*> metrics(Registry::Instance()->GetMetrics());
+  for (auto& m : metrics) {
+    CHECK_NOTNULL(m);
+
+    // See
+    // https://cloud.google.com/monitoring/v2beta2/metricDescriptors#resource
+    // for a description of the structure we're building here.
+
+    JsonArray labels;
+    AddLabelDescription("instance",
+                        "Instance from which the sample originates.", &labels);
+    for (const auto& label : m->LabelNames()) {
+      AddLabelDescription(label, label, &labels);
+    }
+
+    JsonObject desc;
+    switch (m->Type()) {
+      case Metric::COUNTER:
+        // only gauge type metrics are supported for custom metrics currently:
+        // https://cloud.google.com/monitoring/api/metrics#metric-types
+        desc.Add("metricType", "gauge");
+        break;
+      case Metric::GAUGE:
+        desc.Add("metricType", "gauge");
+        break;
+      default:
+        LOG(FATAL) << "Unknown type: " << m->Type();
+    }
+    desc.Add("valueType", "double");
+
+    JsonObject metric;
+    metric.Add("name", kCloudPrefix + m->Name());
+    metric.Add("description", m->Help());
+    metric.Add("labels", labels);
+    metric.Add("typeDescriptor", desc);
+
+    UrlFetcher::Request req(
+        (URL(FLAGS_google_compute_monitoring_metrics_url)));
+    req.verb = UrlFetcher::Verb::POST;
+    req.headers.insert(make_pair("Content-Type", "application/json"));
+    req.headers.insert(make_pair("Authorization", "Bearer " + bearer_token_));
+    req.body = metric.ToString();
+
+    UrlFetcher::Response resp;
+    SyncTask task(executor_);
+    VLOG(1) << "Creating metric m.Name()...";
+    VLOG(2) << req.body;
+    fetcher_->Fetch(req, &resp, task.task());
+    task.Wait();
+    CHECK_EQ(util::Status::OK, task.status());
```

Comment:
Could be handled better in a follow up PR, but if you failed to create the metrics then you're not going to be able to write to them, and your monitoring will have no idea what's going on (despite all the jobs apparently being "up"), seemed better to just bomb out and at least cause all the bells and lights to go off.